dash_core_components/express.py

```diff
@@ -0,0 +1,119 @@
+import dash_html_components as html
+
+from dash.dependencies import Output, Input
+
+
+# region Utils for Download component
+
+
+def send_file(path, filename=None, type=None):
+    """
+    Convert a file into the format expected by the Download component.
+    :param path: path to the file to be sent
+    :param filename: name of the file, if not provided the original filename is used
+    :param type: type of the file (optional, passed to Blob in the javascript layer)
+    :return: dict of file content (base64 encoded) and meta data used by the Download component
+    """
+    # If filename is not set, read it from the path.
+    if filename is None:
+        filename = ntpath.basename(path)
+    # Read the file into a base64 string.
+    with open(path, 'rb') as f:
+        content = base64.b64encode(f.read()).decode()
+    # Wrap in dict.
+    return dict(content=content, filename=filename, type=type, base64=True)
+
+
+def send_bytes(writer, filename, type=None, **kwargs):
+    """
+    Convert data written to BytesIO into the format expected by the Download component.
+    :param writer: a writer that can write to BytesIO
+    :param filename: the name of the file
+    :param type: type of the file (optional, passed to Blob in the javascript layer)
+    :return: dict of data frame content (base64 encoded) and meta data used by the Download component
+    """
+    data_io = io.BytesIO()
+    # Some pandas writers try to close the IO, we do not want that.
+    data_io_close = data_io.close
+    data_io.close = lambda: None
+    # Write data content to base64 string.
+    writer(data_io, **kwargs)
+    data_value = data_io.getvalue()
+    data_io_close()
+    content = base64.b64encode(data_value).decode()
+    # Wrap in dict.
+    return dict(content=content, filename=filename, type=type, base64=True)
+
+
+def send_string(writer, filename, type=None, **kwargs):
+    """
+    Convert data written to StringIO into the format expected by the Download component.
+    :param writer: a writer that can write to StringIO
+    :param filename: the name of the file
+    :param type: type of the file (optional, passed to Blob in the javascript layer)
+    :return: dict of data frame content (base64 encoded) and meta data used by the Download component
+    """
+    data_io = io.StringIO()
+    # Some pandas writers try to close the IO, we do not want that.
+    data_io_close = data_io.close
+    data_io.close = lambda: None
+    # Write data content to base64 string.
+    writer(data_io, **kwargs)
+    data_value = data_io.getvalue().encode()
+    data_io_close()
+    content = base64.b64encode(data_value).decode()
+    # Wrap in dict.
+    return dict(content=content, filename=filename, type=type, base64=True)
+
+
+_known_pandas_writers = {
+    "to_csv": False,
+    "to_json": False,
+    "to_html": False,
+    "to_excel": True,
+    "to_feather": True,
+    "to_parquet": True,
+    "to_msgpack": True,
+    "to_stata": True,
+    "to_pickle": True,
```

Comment:
This `True`/`False` looks funny - as if `False` means "not a known writer"
Perhaps the values could just be `send_bytes`/`send_string`?

That said, are the functions `send_bytes` and `send_string` useful on their own? With `writer` as the first arg it feels pretty specific to pandas, and it doesn't look like a helper (or b64 encoding) is necessary for strings anyway - you just use `dict(content=the_string, filename=the_filename)` right? Does the same apply to byte strings? If that's right and these are only useful internally I'd give them underscored names like `_send_pd_string` - or even inline them in `send_data_frame` and call this dict `_pandas_writes_bytes` - that would be DRYer since the only differences between the two functions are `BytesIO` vs `StringIO` and whether you `.encode()` after `.getvalue()`.

Reply:
Ah, yes, i guess that would actually make more sense :)

Yes, I consider the `send_bytes` and `send_string` functions to be usefull on their own. The writer can be _any_ object, not just pandas writers. The most common use case is probably [sending excel files](https://stackoverflow.com/a/62088521/2428887), but it could be anothing. It's not more than a few weeks ago since i got the latest request on how to send a particular file type, in this case pdf files created by `pdfrw`. Since `pdfrw` creates binary files, the solution is

```
send_bytes(lambda buf: pdfrw.PdfWriter().write(buf, my_pdf_data), filename="my_pdf.pdf")
```

Reply:
Can confirm send_bytes is useful (I was the one who asked about pdfrw).  It's working great within my dash app currently

Reply:
OK, that makes sense. My main concern was that you have to pass a function in, and I imagine in a lot of these situations you'll already have a string or bytestring so it would feel funny to have to wrap that in a `lambda`. What if we allowed that first parameter to be either a writer function or a string / bytestring? For character strings that's only a slight simplification vs constructing the dict yourself, but still may be easier for users to learn. For bytestrings the argument feels more solid, as `send_bytes` ensures we get the encoding right.

Reply:
I see you point. That being said, i haven't yet encountered a case where I have a `bytes` object, passing around a writer seems to be more common (at least in my experience). I have now refactored the code and made some minor changes so that,

* `send_bytes` now accepts a `bytes` object or a writer compatibale with `BytesIO`
* `send_string` now accepts a `str` object or a writer compatiable with `StringIO`

While `send_string` doesn't simplify much as compared to constructing the dictionary manually, i actually like that there are util methods for all common use cases. This way, the data structure (a dict with certain properties) is hidden from the user; it essentially becomes an implementation detail.